File: 121.买卖股票的最佳时机.py
#
# @lc app=leetcode.cn id=121 lang=python3
#
# [121] 买卖股票的最佳时机
#
from common import *
# @lc code=start


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min = float('inf')
        max_profit = 0
        # 找到最小的，然后每天和他做差值
        for i, p in enumerate(prices):
            if p < min:
                min = p
            elif (p - min) > max_profit:
                # max = prices[i]
                # max_i = i
                max_profit = p - min
        return max_profit

# @lc code=end


if __name__ == "__main__":
    s = Solution()
    print(s.maxProfit([7, 1, 5, 3, 6, 4]))
    pass

File: 121.买卖股票的最佳时机.cpp
/*
 * @lc app=leetcode.cn id=121 lang=cpp
 *
 * [121] 买卖股票的最佳时机
 */
#include "commom.hpp"
// @lc code=start
class Solution {
 public:
  int maxProfit(vector<int>& prices) {
    /*
     ! 每天的收益情况有两种:
     * 当天买入了股票，那么我们就是要找负的最大值。

     * 当天卖出了股票，也就是收益，那么我们将历史最大收益和今天卖出的收益进行比较。

    */

    int buy = -prices[0], sell = 0, ans = 0;
    for (int i = 1; i < prices.size(); i++) {
      tie(buy, sell) =
          make_tuple(max(buy, -prices[i]), max(sell, buy + prices[i]));
    }
    return sell;
  }
};
// @lc code=end

int main(int argc, char const* argv[]) {
  Solution s;
  // s.maxProfit();
  return 0;
}
